### PR TITLE
Issue 25852: validate NameConstraints subnet masks

### LINKED ISSUE #25852: OpenSSL converted certificate with incorrect subnet mask
I used the format conversion command of openssl to convert the der certificate to a pem certificate for verification. For my test case, the conversion and verification commands were all executed successfully, but this test case does not comply with RFC5280. In the Nameconstrains extension, a non-compliant mask appears. I think this is a bug.

$ openssl x509 -inform DER -in E_mask.der -out E_mask.pem
$ openssl verify -CAfile E_maskCA.pem E_mask.pem
E_mask.pem: OK

I have provided the test case in the attachment
This test case uses an incorrect subnet mask as the value of Nameconstraints
[Case.zip](https://github.com/user-attachments/files/18268977/Case.zip)

### PR DESCRIPTION
**Pull request by websurfer5**
# Validate NameConstraints Subnet Masks

## Summary

This PR fixes a critical RFC 5280 compliance issue where OpenSSL incorrectly accepts non-contiguous subnet masks in NameConstraints extensions with IP address constraints. The fix adds proper validation to ensure subnet masks are contiguous (CIDR-style) as required by RFC 5280 Section 4.2.1.10, and includes comprehensive tests to verify the fix.

## Related Issue

Fixes #25852 - OpenSSL accepts certificates with non-contiguous subnet masks in NameConstraints extension

## Problem Description

RFC 5280 Section 4.2.1.10 specifies that IP address constraints in NameConstraints extensions MUST use CIDR notation with contiguous subnet masks (as defined in RFC 4632). Prior to this fix, OpenSSL incorrectly accepted non-contiguous subnet masks such as:
- `255.0.255.0` (has gaps)
- `255.255.128.255` (non-contiguous bits)
- `255.255.254.1` (trailing 1-bit after 0-bits)

The existing code in `crypto/x509/v3_ncons.c:849` explicitly documented this issue with a TODO comment:
```c
/* Not checking for wrong mask definition: i.e.: 255.0.255.0 */
```

This allowed certificates with invalid NameConstraints to be created and accepted, violating RFC 5280 requirements and potentially enabling security issues.

### Core Fix

**`crypto/x509/v3_ncons.c`**

1. **New function `is_valid_netmask()`**:
   - Validates that subnet masks have contiguous 1-bits followed by contiguous 0-bits
   - Supports both IPv4 (4 bytes) and IPv6 (16 bytes)
   - Checks byte-by-byte and bit-by-bit for proper CIDR format
   - Returns 0 for invalid masks, 1 for valid masks

2. **Updated `nc_ip()` function**:
   - Calls `is_valid_netmask()` before processing IP constraints
   - Returns `X509_V_ERR_UNSUPPORTED_NAME_SYNTAX` for invalid masks
   - Removed the TODO comment acknowledging the missing validation

### Test Suite

**`test/nc_invalid_mask_test.c`**
- C API test program with 6 comprehensive test cases
- Tests both IPv4 and IPv6 invalid and valid subnet masks
- Uses public OpenSSL APIs (X509V3_EXT_conf_nid, EVP_PKEY_*, X509_*)
- Creates CA certificates with NameConstraints and validates behavior

**`test/recipes/25-test_verify_nc_invalid_mask.t`** (194 lines)
- Integrated test recipe with 17 total tests
- **Part 1** (Test 1): Runs C API test program
- **Part 2** (Tests 2-17): CLI-based certificate creation and verification
  - Tests 2-10: Verify invalid masks are properly rejected
  - Tests 11-17: Verify valid masks continue to work correctly

### Validation Logic

The `is_valid_netmask()` function validates subnet masks by:

1. **Length check**: Ensures mask is exactly 4 bytes (IPv4) or 16 bytes (IPv6)
2. **Contiguity check**: Iterates through all bytes
   - Tracks whether any 0-bits have been seen (`found_zero`)
   - For full bytes: `0xff` allowed before any `0x00`, `0x00` sets `found_zero`
   - For partial bytes: Checks each bit, ensuring no 1-bits appear after 0-bits
3. **Returns**: 1 if mask is valid (contiguous), 0 otherwise

### Error Handling

Invalid masks now return `X509_V_ERR_UNSUPPORTED_NAME_SYNTAX` which will:
- Cause certificate creation with invalid NameConstraints to fail
- Cause certificate verification to fail if CA has invalid NameConstraints
- Provide clear error message to users

### Invalid Subnet Masks (Now Properly Rejected)

**IPv4 Test Cases**:
- `255.0.255.0` - Gap in the middle (bytes: 0xff, 0x00, 0xff, 0x00)
- `255.255.128.255` - Non-contiguous bits (bytes: 0xff, 0xff, 0x80, 0xff)
- `255.255.254.1` - Trailing 1-bit after 0-bits (bytes: 0xff, 0xff, 0xfe, 0x01)

**IPv6 Test Cases**:
- `ffff:ffff:ffff:ffff:0000:0000:ffff:ffff` - Gap in the middle

### Valid Subnet Masks (Continue to Work)

**IPv4 Valid Masks**:
- `255.255.255.255` (/32) - Host mask
- `255.255.255.0` (/24) - Class C network
- `255.255.0.0` (/16) - Class B network
- `255.0.0.0` (/8) - Class A network
- `255.255.255.252` (/30) - Point-to-point link
- `255.255.255.128` (/25) - Subnet

**IPv6 Valid Masks**:
- `ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff` (/128) - Host
- `ffff:ffff:ffff:ffff:0000:0000:0000:0000` (/64) - Subnet
- `ffff:ffff:ffff:ffff:ffff:ffff:0000:0000` (/96) - Prefix

### CODE REVIEW

File: crypto/x509/v3_ncons.c [hunk 1]
```diff
@@ -824,6 +825,46 @@ static int nc_uri(ASN1_IA5STRING *uri, ASN1_IA5STRING *base)
     return ret;
 }
 
+static int is_valid_netmask(unsigned char *mask, int len)
```
Comment (shahsb) [inline crypto/x509/v3_ncons.c:846]: Add a brief comment explaining the contiguity check (e.g., "RFC 5280 §4.2.1.10 requires CIDR-style masks"). No deletions, so low regression risk.

File: crypto/x509/v3_ncons.c [hunk 2]
```diff
@@ -845,8 +886,10 @@ static int nc_ip(ASN1_OCTET_STRING *ip, ASN1_OCTET_STRING *base)
 
     maskptr = base->data + hostlen;
 
+    if (!is_valid_netmask(maskptr, hostlen))
```
Comment (shahsb) [inline crypto/x509/v3_ncons.c]: **Documentation:** Add a blurb to doc/man3/X509_VERIFY_PARAM_set_flags.pod or CHANGES.md (e.g., "3.3.1: Reject non-contiguous IP subnet masks in NameConstraints per RFC 5280").

File: test/nc_invalid_mask_test.c [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+}
+
+int setup_tests(void)
+{
+    ADD_TEST(test_ipv4_noncontiguous_mask_1);
+    ADD_TEST(test_ipv4_noncontiguous_mask_2);
+    ADD_TEST(test_ipv4_noncontiguous_mask_3);
+    ADD_TEST(test_ipv4_valid_masks);
+    ADD_TEST(test_ipv6_noncontiguous_mask);
+    ADD_TEST(test_ipv6_valid_masks);
```
Comment (shahsb) [inline test/nc_invalid_mask_test.c:656]: Add a fuzz target for random masks to catch any overlooked edge cases (e.g., via libFuzzer integration if not already present).

**Review by shahsb — COMMENTED** [review]
No performance impact noted, but validation adds a bit-level scan per constraint -- negligible for typical cert sizes.

**Comment by websurfer5** [comment]
I posted before it was ready.  It works correctly now.  I added the recommended changes.
